### ISSUE TITLE
Erroneous output in scenario containing NamedTuple, TypeVar, Protocol, and Generic

### ISSUE THREAD

**Issue by coder0xff** [python/mypy#9357]
* Are you reporting a bug, or opening a feature request?
Bug

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```python
from typing import Callable, Generic, NamedTuple, Optional, Protocol, TypeVar


GenericType = TypeVar("GenericType")


class ClassOne(Protocol[GenericType]):
    def function(self, _: GenericType) -> Optional[GenericType]:
        return None


class ClassTwo(NamedTuple, Generic[GenericType]):
    not_: Callable[[ClassOne[GenericType]], Optional[ClassOne[GenericType]]]


def foobar(_: ClassOne[GenericType]):
    def function(_: ClassOne[GenericType]) -> Optional[ClassOne[GenericType]]:
        return None

    return ClassTwo(function)

```

* What is the actual behavior/output?

    Argument 1 to "ClassTwo" has incompatible type "Callable[[Arg(ClassOne[GenericType], '_')], Optional[ClassOne[GenericType]]]"; expected "Callable[[ClassOne[GenericType]], Optional[ClassOne[GenericType]]]"

* What is the behavior/output you expect?
It should not produce any error

* What are the versions of mypy and Python you are using?
Python 3.8.5
mypy 0.782

* Do you see the same issue after installing mypy from Git master?
Yes

* What are the mypy flags you are using? (For example --strict-optional)
None

* If mypy crashed with a traceback, please paste the full traceback below.
N/A

# NOTES
* The only difference in the types is the "Arg" part, which should be irrelevant.
* I was not able to remove any element of this reproduction and still produce the problem.

**Comment by gvanrossum** [comment]
If you rename the argument to `function` from `_` to `__unused`, the `Arg` part of the message disappears, so you're right that that's not the problem.

Here's a simpler demo showing that the `Protocol` is also a red herring (as is `Optional`):
```py
from typing import Callable, Generic, NamedTuple, TypeVar

T = TypeVar("T")
S = TypeVar("S")

class C(NamedTuple, Generic[T]):
    x: Callable[[T], T]

def foobar(arg: S) -> None:
    def f2(__unused: S) -> S:
        ...

    C(f2)  # Error here
```
which gives the same error:
```
u.py:13: error: Argument 1 to "C" has incompatible type "Callable[[S], S]"; expected "Callable[[T], T]"
```
I'm guessing that something is wrong with `NamedTuple`, so I'm labeling it as such. I'm still not sure it's a bug -- it could also be that `foobar` being a function generic in `S` makes it just different enough. Adding this to the end of the file is perhaps revealing:
```py
    reveal_type(C)
    reveal_type(f2)
```
Output:
```
u.py:14: note: Revealed type is 'def (x: def (T`1) -> T`1) -> Tuple[def (T`1) -> T`1, fallback=u.C]'
u.py:15: note: Revealed type is 'def (S`-1) -> S`-1'
```

**Comment by hauntsaninja** [comment]
Hmm, from the reveal_type it looks like the named tuple isn't actually generic.

Oh, this might be what's going on:
```
~/tmp λ cat test51.py
from typing import *

T = TypeVar("T")

class C(NamedTuple, Generic[T]):
    x: Callable[[T], T]

x: C[int]
~/tmp λ mypy test51.py
test51.py:8: error: "C" expects no type arguments, but 1 given
test51.py:8: error: Generic tuple types not supported
Found 2 errors in 1 file (checked 1 source file)
```

**Comment by gvanrossum** [comment]
So that means it’s (yet another) NamedTuple bug right?

**Comment by erictraut** [comment]
This bug appears to have been fixed. I can not repro it with the latest version of mypy.

**Comment by hauntsaninja** [comment]
Yeah, I think we got generic namedtuples in 0.990 or something
